Port `kbn-pm` to TypeScript.

azasypkin opened the pull request:
Changes:
* `kbn-pm` is ported to `TypeScript`
* Some `kbn-pm` dependencies are bumped up

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/15796/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/15800/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/15808/)

kjbekkelund commented on packages/kbn-build/package.json:
```diff
@@ -5,36 +5,60 @@
   "scripts": {
-    "build": "webpack"
+    "build": "webpack",
+    "prettier": "prettier --single-quote --write './src/**/*.ts'"
```
Hm, I wonder if it's possible to handle this in the ESLint setup, https://github.com/elastic/kibana/blob/master/.eslintrc.js#L17. Maaaybe an additional override that specifies `parser` as https://github.com/eslint/typescript-eslint-parser?

azasypkin replied:
I didn't manage to make it work from the first try, but will try again :)

kjbekkelund commented on packages/kbn-build/src/commands/clean.ts:
```diff
@@ -28,7 +33,8 @@ export async function run(projects, projectGraph, { rootPath }) {
     for (const dir of directoriesToDelete) {
       const deleting = del(dir, { force: true });
-      ora.promise(deleting, relative(rootPath, dir));
+      // @types/ora doesn't expose `ora.promise`.
```
:( Maybe PR a fix to DefinitelyTyped?

azasypkin replied:
Yes, the plan is to augment/workaround everything we need, merge (if everything is fine) and then contribute to DefinitelyTyped and get rid of these in our codebase. The reason: PRs to DT may take some time. What do you think?

kjbekkelund replied:
Works for me :+1:

azasypkin replied:
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/23699

kjbekkelund replied:
🎉

azasypkin replied:
:heavy_check_mark:

kjbekkelund commented on packages/kbn-build/src/commands/run.ts:
```diff
@@ -1,13 +1,18 @@
 
-export async function run(projects, projectGraph, { extraArgs }) {
+export async function run(
+  projects: Map<string, Project>,
```
Maybe create a type that's used both in the commands and in `packages/kbn-build/src/commands/index.ts`?

kjbekkelund replied:
(Could use generics for the `options` arg)

kjbekkelund commented on packages/kbn-build/src/config.ts:
```diff
@@ -3,7 +3,10 @@ import { resolve } from 'path';
-export function getProjectPaths(rootPath, options) {
+export function getProjectPaths(
+  rootPath: string,
+  options: { [key: string]: any }
```
Would be nice to type the options, maybe something along the lines of

```
type ProjectPathOptions = {
  'skip-kibana-extra': boolean,
  'skip-kibana': boolean,
  [key: string]: any
}
```

(haven't tried to make that change, so don't know if it makes sense across the code base)

azasypkin replied:
Yeah, I was thinking about that too, but maybe we should omit `[key: string]: any`part and be stricter?

kjbekkelund replied:
++ that would be great

azasypkin replied:
:heavy_check_mark:

azasypkin commented on packages/kbn-build/src/commands/bootstrap.test.ts:
```diff
@@ -1,24 +1,26 @@
+const installInDir = jest.fn();
+const runScriptInPackageStreaming = jest.fn();
+const linkProjectExecutables = jest.fn();
```
> Also, Jest didn't complain about missing mock prefix? Strange. I

Hmm, haven't seen any error/warning on that 🤔 

> Maybe that's something worth exploring?

Let me check if/how I can apply that in this case :)

azasypkin replied:
:heavy_check_mark: (at least how I understood it) :)

azasypkin commented on packages/kbn-build/src/commands/bootstrap.test.ts:
```diff
@@ -75,7 +77,10 @@ test('handles dependencies of dependencies', async () => {
   const projectGraph = buildProjectGraph(projects);
 
-  const logMock = jest.spyOn(console, 'log').mockImplementation(noop);
+  // Workaround for https://github.com/DefinitelyTyped/DefinitelyTyped/issues/17605.
```
I'll try and see if my `.d.ts-fu` is enough to do that :)

azasypkin replied:
Okay, I see at least two possibilities:
1. Complement [`MockInstance`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jest/index.d.ts#L528) with two function signatures that [`Mock`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jest/index.d.ts#L505) has and get rid of the latter. And update [the relevant part of interface](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jest/index.d.ts#L532-L537) to something like this instead:
```ts
mockImplementation(fn: (...args: any[]) => any): this;
mockImplementationOnce(fn: (...args: any[]) => any): this;
mockReturnThis(): this;
mockReturnValue(value: any): this;
mockReturnValueOnce(value: any): this;
mockName(name: string): this;
```
That looks like correct solution to me, but I need to read Jest code to know for sure. The downside is that it's not easy/possible to implement this solution via augmentation, so we'll have to live with `const mock: jest.SpyInstance<any> = jest.spyOn(...).mockImplementation(...) as any;` workaround all over the place.

2. Make `SpyInstance` a "sub-interface" of `Mock`, and override some methods specifically:
```ts
interface SpyInstance<T> extends Mock<T> {
    mockImplementation(fn: (...args: any[]) => any): SpyInstance<T>;
    mockImplementationOnce(fn: (...args: any[]) => any): SpyInstance<T>;
    mockReturnThis(): SpyInstance<T>;
    mockReturnValue(value: any): SpyInstance<T>;
    mockReturnValueOnce(value: any): SpyInstance<T>;
    mockName(name: string): SpyInstance<T>;
  }
}
```
The benefit here is that we can patch Jest declarations locally and unblock ourselves. And at the same time propose a proper fix (if any of these is proper one :)) to DefinitelyTyped.

What would you prefer?

kjbekkelund replied:
I’m leaning towards solution 2

azasypkin replied:
👍

azasypkin replied:
:heavy_check_mark: (added our custom augmentation, will prepare PR to upstream Jest typings soon)

azasypkin commented on packages/kbn-build/src/commands/run.ts:
```diff
@@ -26,9 +31,9 @@ export async function run(projects, projectGraph, { extraArgs }) {
   );
 
-  await parallelizeBatches(batchedProjects, pkg => {
+  await parallelizeBatches(batchedProjects, async (pkg: Project) => {
```
Hmm, right, for some reason I've added it before `parallelizeBatches` has been extracted and left after rebase, but theoretically it shouldn't matter... Will remote it, thanks.

azasypkin replied:
:heavy_check_mark:

azasypkin commented on packages/kbn-build/types/getopts/index.d.ts:
```diff
@@ -0,0 +1,10 @@
+declare module 'getopts' {
```
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/23707

azasypkin replied:
:heavy_check_mark: (merged)

azasypkin commented on packages/kbn-build/types/write_pkg/index.d.ts:
```diff
@@ -0,0 +1,6 @@
+declare module 'write-pkg' {
```
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/23703

azasypkin replied:
:heavy_check_mark: (merged)

azasypkin commented on packages/kbn-build/types/strong_log_transformer/index.d.ts:
```diff
@@ -0,0 +1,11 @@
+declare module 'strong-log-transformer' {
```
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/23711

azasypkin replied:
:heavy_check_mark: (merged)

azasypkin commented:
@kjbekkelund I've rebased PR on the latest master and fixed all the review comments. Still trying to integrate eslint, but apart from that everything is ready for review.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/52/)

kjbekkelund commented on packages/kbn-pm/src/utils/package_json.ts:
```diff
@@ -2,20 +2,25 @@ import readPkg from 'read-pkg';
 import path from 'path';
 
-export function readPackageJson(dir) {
+export type Package = { [k: string]: any };
```
`Package` -> `PackageJson`?

azasypkin replied:
:heavy_check_mark:

kjbekkelund commented on packages/kbn-pm/src/commands/bootstrap.test.ts:
```diff
@@ -18,7 +13,11 @@ import { buildProjectGraph } from '../utils/projects';
+const mockRunScriptInPackageStreaming: jest.Mock = runScriptInPackageStreaming as any;
+const mockLinkProjectExecutables: jest.Mock = linkProjectExecutables as any;
+
+const createProject = (fields: { [key: string]: any }, path = '.') =>
```
`fields: Package`? (using the `Package` type in `utils/package_json`)

kjbekkelund commented on tsconfig.json:
```diff
@@ -0,0 +1,48 @@
+{
+  "compileOnSave": true,
```
We can remove this. I added it unnecessarily to the `new-platform` and have forgotten about it.

azasypkin replied:
:heavy_check_mark:

kjbekkelund commented on tsconfig.json:
```diff
@@ -0,0 +1,48 @@
+{
+  "compileOnSave": true,
+  "compilerOptions": {
+    "stripInternal": true,
```
We aren't generating declaration files yet, so I think we can remove this too(?)

azasypkin replied:
:heavy_check_mark:

kjbekkelund commented on tsconfig.json:
```diff
@@ -0,0 +1,48 @@
+    "module": "esnext",
+
+    // Allows default imports from modules with no default export. This does not affect code emit, just type checking.
+    "allowSyntheticDefaultImports": true,
```
I think we can remove this, as it should be handled automatically by the `esModuleInterop` flag.

azasypkin replied:
It should, but it doesn't. Let me investigate, maybe I'm doing something wrong.

azasypkin replied:
Here is what I see in TS source code:
```ts
function getAllowSyntheticDefaultImports(compilerOptions) {
        var moduleKind = getEmitModuleKind(compilerOptions);
        return compilerOptions.allowSyntheticDefaultImports !== undefined
            ? compilerOptions.allowSyntheticDefaultImports
            : compilerOptions.esModuleInterop
                ? moduleKind !== ts.ModuleKind.None && moduleKind < ts.ModuleKind.ES2015
                : moduleKind === ts.ModuleKind.System;
    }
```

kjbekkelund replied:
Ah, fun. Okey, let's keep it. Want to add a note about that in the comment?

kjbekkelund commented on tsconfig.json:
```diff
@@ -0,0 +1,48 @@
+    "noImplicitReturns": false,
+
+    // Report errors on unused locals.
+    "noUnusedLocals": true,
```
I know this is `true` in `new-platform`, but I think we should turn it off and rather focus on improving the lint setup. This is annoying when just exploring things in code, but not commiting yet.

kjbekkelund replied:
I think we can also just remove the `noImplicitReturns` flag from here. (It's also `false` by default)

azasypkin replied:
Yeah it (`noUnusedLocals` and `noImplicitReturns`) feels like linter responsibility, my only minor concern is that we don't have linter setup yet. On the other hand it shouldn't do any harm, so I'll remove these settings.

kjbekkelund commented on src/dev/jest/tsconfig.json:
```diff
@@ -0,0 +1,17 @@
+    // update the target to a newer version of EcmaScript.
+    "target": "es2015"
+  },
+  "include": [
```
Did some testing, looks like we maybe can do this instead:

```diff
diff --git a/src/dev/jest/tsconfig.json b/src/dev/jest/tsconfig.json
index e19a53a81e..0bfce29894 100644
--- a/src/dev/jest/tsconfig.json
+++ b/src/dev/jest/tsconfig.json
@@ -12,6 +12,6 @@
     "target": "es2015"
   },
   "include": [
-    "../../../packages/kbn-pm/**/*"
+    "../../../**/*.ts"
   ]
 }
diff --git a/tsconfig.json b/tsconfig.json
index afece44dff..76871cfb27 100644
--- a/tsconfig.json
+++ b/tsconfig.json
@@ -43,6 +43,6 @@
   },
 
   "exclude": [
-    "node_modules"
+    "**/node_modules/**"
   ]
 }
```

What do you think?

azasypkin replied:
`"../../../**/*.ts"` ++

What is the use case where `node_modules` doesn't work, but `"**/node_modules/**"` works?

kjbekkelund replied:
It looked like it only included the top-level `node_modules`, not _all_ node_modules in the tree (e.g. within `packages`). I didn't look deeper into that though, so might be mistaken.

azasypkin replied:
I believe it excludes every `node_modules`. Just to double check I've experimented with _bad_ file inside `src/node_modules` and it worked as expected: file has been skipped. Actually TS excludes `node_modules` by default (http://www.typescriptlang.org/docs/handbook/tsconfig-json.html), but I still prefer to mention that explicitly.

kjbekkelund replied:
:tada: That's great. It was just the TS output that confused me, thanks for checking.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/58/)

azasypkin commented on packages/kbn-pm/src/utils/projects.ts:
```diff
@@ -7,7 +7,13 @@ import { Project } from './project';
 
 const glob = promisify(_glob);
```
pify :man_facepalming: .

azasypkin replied:
:heavy_check_mark: (switched to `bluebird` as we agreed on Slack).

azasypkin commented on packages/kbn-pm/dist/index.js:
```diff
@@ -61,7 +61,7 @@ module.exports =
 /******/ 	__webpack_require__.p = "";
```
__note:__ Bluebird added ~4000 lines to `dist/index.js` :)

kjbekkelund replied:
Shit happens. Hopefully we'll move to Node8 soon.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/64/)

azasypkin commented:
Looks like CI was hit by the network issue "An unexpected error occurred: "https://registry.yarnpkg.com/@types/lodash/-/lodash-4.14.104.tgz: unexpected end of file".", re-trigger build

Jenkins, test this

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/66/)

azasypkin commented:
6.x/6.3: 8e97b7d2850638edf4ca0ce27390de274f869805
